src/Spice86.Core/Emulator/Devices/Sound/PCSpeaker/PcSpeaker.cs

```diff
@@ -1,66 +1,75 @@
 
     private readonly int _outputSampleRate = 48000;
     private readonly int _ticksPerSample;
-    private readonly LatchedUInt16 _frequencyRegister;
+    private readonly Pit8254Counter _pit8254Counter;
     private readonly Stopwatch _durationTimer = new();
-    private readonly ConcurrentQueue<QueuedNote> _queuedNotes = new();
-    private readonly object _threadStateLock = new();
+    private QueuedNote _currentNote;
     private SpeakerControl _controlRegister = SpeakerControl.UseTimer;
-    private Task? _generateWaveformTask;
-    private readonly CancellationTokenSource _cancelGenerateWaveform = new();
     private int _currentPeriod;
 
+    private readonly Thread _playbackThread;
+    private bool _playbackStarted;
+    private volatile bool _endPlayback;
+    private readonly IPauseHandler _pauseHandler;
+
     private bool _disposed;
 
     /// <summary>
     /// Initializes a new instance of <see cref="PcSpeaker"/>
     /// </summary>
     /// <param name="softwareMixer">The software mixer for sound channels.</param>
+    /// <param name="pit8254Counter">PIT8254 counter backing up the PC speaker.</param>
     /// <param name="state">The CPU registers and flags.</param>
     /// <param name="ioPortDispatcher">The class that is responsible for dispatching ports reads and writes to classes that respond to them.</param>
+    /// <param name="pauseHandler">The handler for the emulation pause state.</param>
     /// <param name="loggerService">The logger service implementation.</param>
     /// <param name="failOnUnhandledPort">Whether we throw an exception when an I/O port wasn't handled.</param>
-    public PcSpeaker(SoftwareMixer softwareMixer, State state, IOPortDispatcher ioPortDispatcher,
-        ILoggerService loggerService, bool failOnUnhandledPort) : base(state, failOnUnhandledPort, loggerService) {
+    public PcSpeaker(SoftwareMixer softwareMixer,
+        State state,
+        Pit8254Counter pit8254Counter,
+        IOPortDispatcher ioPortDispatcher,
+        IPauseHandler pauseHandler,
+        ILoggerService loggerService,
+        bool failOnUnhandledPort) : base(state, failOnUnhandledPort, loggerService) {
         _soundChannel = softwareMixer.CreateChannel(nameof(PcSpeaker));
-        _frequencyRegister = new();
-        _frequencyRegister.ValueChanged += FrequencyChanged;
+        _pit8254Counter = pit8254Counter;
+        pit8254Counter.SettingChangedEvent += OnTimerSettingChanged;
         _ticksPerSample = (int)(Stopwatch.Frequency / (double)_outputSampleRate);
         InitPortHandlers(ioPortDispatcher);
+        _pauseHandler = pauseHandler;
+        _playbackThread = new Thread(AudioPlayback) {
+            Name = nameof(PcSpeaker),
+        };
     }
 
     /// <summary>
     /// Gets the current frequency in Hz.
     /// </summary>
-    private double Frequency => FrequencyFactor / _frequencyRegister;
+    private double Frequency => _pit8254Counter.Activator.Frequency;
```

Comment:
The class talks to 'a friend of a friend' here.

should it not directly depend on _activator.Frequency ?